crates/revm_precompiles/src/fhe.rs

```diff
@@ -133,15 +199,55 @@ where
     let ix_1: usize = u32::from_be_bytes(ix_1.try_into().map_err(|_| FheErr::UnexpectedEOF)?)
         .try_into()
         .map_err(|_| FheErr::PlatformArchitecture)?;
-    let ix_2: usize = u32::from_be_bytes(ix_2.try_into().unwrap())
+    let ix_2: usize = u32::from_be_bytes(ix_2.try_into().map_err(|_| FheErr::UnexpectedEOF)?)
         .try_into()
         .map_err(|_| FheErr::PlatformArchitecture)?;
 
     let pubk = bincode::deserialize(&input[8..ix_1]).map_err(|_| FheErr::InvalidEncoding)?;
     let arg1 = bincode::deserialize(&input[ix_1..ix_2]).map_err(|_| FheErr::InvalidEncoding)?;
     let arg2 = bincode::deserialize(&input[ix_2..]).map_err(|_| FheErr::InvalidEncoding)?;
 
-    let result = op(pubk, arg1, arg2).unwrap();
+    let result = op(arg1, arg2, pubk).unwrap();
+
+    Ok(PrecompileOutput::without_logs(
+        op_cost,
+        bincode::serialize(&result).unwrap(),
+    ))
+}
+
+/// Expects `input` with encoding (not to scale!)
+///
+/// ```text
+/// <---- 4B ----><~---------------~><- 8B ->
+/// [Arg 1 offset][Public key][Arg 1][Arg 2 ]
+/// ```
+///
+/// where argument 1 is a ciphertext, argument 2 is a plaintext u64, and both
+/// the ciphertext and public key are bincode encoded. The resulting returned
+/// ciphertext is bincode encoded.
+fn fhe_binary_op_plain<F>(op_cost: u64, op: F, input: &[u8], gas_limit: u64) -> PrecompileResult
+where
+    F: FnOnce(Ciphertext, Signed, PublicKey) -> Result<Ciphertext, RuntimeError>,
+{
+    if op_cost > gas_limit {
+        return Err(Error::OutOfGas);
+    }
+    if input.len() < 12 {
+        return Err(FheErr::UnexpectedEOF.into());
+    }
+    let ix = &input[..4];
+    let ix: usize = u32::from_be_bytes(ix.try_into().map_err(|_| FheErr::UnexpectedEOF)?)
+        .try_into()
+        .map_err(|_| FheErr::PlatformArchitecture)?;
```

Comment:
This conversion into a usize happens a few times with the same call/errors, potentially abstract out. Would alleviate missing one case like in `fhe_binary_op`

Reply:
Actually the `ix.try_into()` is safe to unwrap given the length check above, so we can even remove that part. Cleans this up a little bit.